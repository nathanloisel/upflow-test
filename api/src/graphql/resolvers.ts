import { ApolloError, IResolvers } from "apollo-server-express";
import ParaCompDb from "../data-sources/db";
import { ParaglindingPractices, CENCertifications } from "../types";

export type SortByField = 'name' | 'brand' | 'allon' | 'CEN' | 'practice' | 'ratio';

interface IGetParaglindersRequest {
  pageSize: number,
  pageNumber: number,
  sortBy?: SortByField,
}

interface IUpdateParaglinderRequest {
  name: string,
  brand: string,
  allon?: number,
  CEN?: CENCertifications,
  practice?: ParaglindingPractices,
  ratio?: number
}

interface IDeleteParaglinderRequest {
  name: string,
  brand: string,
}


type ParaglinderResolvers = IResolvers<any, { dataSources: { db: ParaCompDb } }>;

const resolvers: ParaglinderResolvers = {
  Query: {
    paraglinders: async (_, { pageSize = 20, pageNumber = 1, sortBy }: IGetParaglindersRequest, { dataSources }) => {
      try {
        // parameters validation
        if (pageSize < 1)
          throw new ApolloError("Invalid pageSize parameter. Should be >= 1");
        if (pageNumber < 1)
          throw new ApolloError("Invalid pageNumber parameter. Should be >= 1");

        // get paraglinders entities.
        // add order by name if asked and pagination
        const paraglinders = await dataSources.db.getParaglinders(pageSize, pageNumber, sortBy);

        const [{ count: total }] = await dataSources.db.getTotal();

        return {
          paraglinders,
          pagination: {
            // total number of item
            total,
            // total number of page
            pageCount: Math.floor((total - 1) / pageSize) + 1,
            // current page number
            currentPage: pageNumber
          }
        }
      } catch (e) {
        // the error need to be logged
        throw new ApolloError("Internal Server Error");
      }
    }
  },
  Mutation: {
    updateParaglinder: async (_, { name, brand, CEN, allon, practice, ratio }: IUpdateParaglinderRequest, { dataSources }) => {
      try {
        await dataSources.db.updateParaglinder(name, brand, CEN, allon, practice, ratio);
        return dataSources.db.getParaglinder(name, brand);
      } catch (e) {
        throw new ApolloError(e);
      }
    },

    deleteParaglinder: async (_, { name, brand }: IDeleteParaglinderRequest, { dataSources }) => {
      try {
        return await dataSources.db.deleteParaglinder(name, brand) === 1;
      } catch (e) {
        throw new ApolloError(e);
      }
    }
  }
};

export default resolvers;

