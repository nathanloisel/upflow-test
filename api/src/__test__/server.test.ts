import { ApolloServer } from 'apollo-server-express';

import ParaCompDb from '../data-sources/db';
import typeDefs from '../graphql/type-defs';
import resolvers from '../graphql/resolvers';

const GET_PARAGLINDERS = `
  query GetParaglinders {
    paraglinders {
      paraglinders {
        name
        brand
        allon
        practice
        ratio
      }
      pagination {
        total
        pageCount
        currentPage
      }
    }
  }
`;

describe("Integration test against apollo server", () => {

  const db = new ParaCompDb();
  const apolloServer = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      db
    }),
  });

  // @ts-ignore
  db.getTotal = jest.fn(() => Promise.resolve([{ count: 3 }]));

  // @ts-ignore
  db.getParaglinders = jest.fn(() => Promise.resolve([
    {
      "name": "9cb3da02",
      "brand": "Nitracyr",
      "allon": 4.77,
      "CEN": "B",
      "practice": "ecole",
      "ratio": 11
    },
    {
      "name": "5b96396b",
      "brand": "Avit",
      "allon": 6.07,
      "CEN": "A",
      "practice": "hikeNfly",
      "ratio": 10
    },
    {
      "name": "94be658d",
      "brand": "Avit",
      "allon": 5.6,
      "CEN": "A",
      "practice": "ecole",
      "ratio": 11
    }]));

  it("test apollo server return paraglinders", async () => {

    // run query against the server and snapshot the output
    const res = await apolloServer.executeOperation({ query: GET_PARAGLINDERS });
    expect(res).toMatchSnapshot();
  });

});