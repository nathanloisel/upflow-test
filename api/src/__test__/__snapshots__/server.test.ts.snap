// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test against apollo server test apollo server return paraglinders 1`] = `
Object {
  "data": Object {
    "paraglinders": Object {
      "pagination": Object {
        "currentPage": 1,
        "pageCount": 1,
        "total": 3,
      },
      "paraglinders": Array [
        Object {
          "allon": 4.77,
          "brand": "Nitracyr",
          "name": "9cb3da02",
          "practice": "ecole",
          "ratio": 11,
        },
        Object {
          "allon": 6.07,
          "brand": "Avit",
          "name": "5b96396b",
          "practice": "hikeNfly",
          "ratio": 10,
        },
        Object {
          "allon": 5.6,
          "brand": "Avit",
          "name": "94be658d",
          "practice": "ecole",
          "ratio": 11,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
